### ISSUE TITLE
Error while running `predict_with_paths`

### ISSUE THREAD

**Issue by kategerasimenko** [machamp-nlp/machamp#21]
I am trying to run `predict_with_paths` on the dataset with labels. Call [at the end of training](https://github.com/machamp-nlp/machamp/blob/master/machamp/model/trainer.py#L272) and my custom call fail with the following error. Intermediate evaluation with the same metric during training runs without failing though.

```
│ ~/machamp_repo/machamp/predictor/predict.py:261 in                           │
│ predict_with_paths                                                           │
│                                                                              │
│   259 │                                                                      │
│   260 │   if not raw_text:                                                   │
│ ❱ 261 │   │   metrics = model.get_metrics()                                  │
│   262 │   │   report_metrics(metrics)                                        │
│   263 │   │   eval_file = output_path + '.eval'                              │
│   264 │   │   json.dump(metrics, open(eval_file, 'w'), indent=4)             │
│                                                                              │
│ ~/machamp_repo/machamp/model/machamp.py:442 in                               │
│ get_metrics                                                                  │
│                                                                              │
│   439 │   │   sum_metrics = 0                                                │
│   440 │   │   for decoder in self.decoders:                                  │
│ ❱ 441 │   │   │   names, scores, types = self.decoders[decoder].get_metrics( │
│   442 │   │   │   for name, score, metric_type in zip(names, scores, types): │
│   443 │   │   │   │   name = decoder + '-' + name                            │
│   444 │   │   │   │   metrics[name] = score                                  │
│                                                                              │
│ ~/machamp_repo/machamp/model/machamp_decoder.py:22                           │
│ in get_metrics                                                               │
│                                                                              │
│   19 │   │   self.metric.reset()                                             │
│   20 │                                                                       │
│   21 │   def get_metrics(self):                                              │
│ ❱ 22 │   │   return self.metric.get_scores()                                 │
│   23                                                                         │
│                                                                              │
│ ~/machamp_repo/machamp/metrics/metric.py:76 in                               │
│ get_scores                                                                   │
│                                                                              │
│   73 │   │   scores = []                                                     │
│   74 │   │   types = []                                                      │
│   75 │   │   for metric in self.metrics:                                     │
│ ❱ 76 │   │   │   name, score = self.metrics[metric].get_score()              │
│   77 │   │   │   names.append(name)                                          │
│   78 │   │   │   scores.append(score)                                        │
│   79 │   │   │   types.append(type(self.metrics[metric]))                    │
│                                                                              │
│ ~/machamp_repo/machamp/metrics/f1.py:72 in                                   │
│ get_score                                                                    │
│                                                                              │
│   69 │   │   │   f1s = []                                                    │
│   70 │   │   │   for label_idx in range(1, len(self.tps)):                   │
│   71 │   │   │   │   f1s.append(self.get_f1(self.tps[label_idx], self.fps[la │
│ ❱ 72 │   │   │   return self.str, sum(f1s) / len(f1s)                        │
│   73 │   │                                                                   │
│   74 │   │   elif self.type_f1 == 'binary':                                  │
│   75 │   │   │   if len(self.tps) > 3:                                       │
╰──────────────────────────────────────────────────────────────────────────────╯
ZeroDivisionError: division by zero
```
(line numbers can differ from this repository because I included a couple of custom features).

**Comment by robvanderg** [comment]
Thanks for reporting the issue, I expect to have time to look into this next week

**Comment by robvanderg** [comment]
It seems to be doing predictions without calculating a metric (did you supply gold data?), but then it somehow tried to evaluate. Simply checking if len(f1) == 0 and returning 0.0 should resolve this. I will push this in a minute.

**Comment by kategerasimenko** [comment]
I supplied the gold data, I just ran the training with the same dev dataset. Metric during training was calculated without any issues but evaluation afterwards on the same dev dataset failed. If you cannot reproduce this error, can try to reproduce this with some open-source datasets and send the code here next week :)

**Comment by kategerasimenko** [comment]
Hello! Sorry for the delayed reply, here's the [colab notebook](https://colab.research.google.com/drive/1CQz2fjbInhSQjEYeRa5YB8htfbu80sJx?usp=sharing) with both issues. 

For this issue, I made one further small fix and there're no exceptions now, but when running predict for each dataset, the metric runs for all datasets, giving 0.0 for other datasets. See `.eval` files in the outputs:

trec.out.eval:
```
{
    "trec-f1_macro": 0.9672453724685518,
    "tweets-f1_macro": 0.0,
    "sum": 0.9672453724685518
}
```

tweets.out.eval:
```
{
    "trec-f1_macro": 0.0,
    "tweets-f1_macro": 0.7858704141646664,
    "sum": 0.7858704141646664
}
```

As a user, I would expect only one score for every dataset or one joint `.eval` file.

**Comment by robvanderg** [comment]
> For this issue, I made one further small fix  
This refers to the return statement in F1 score I assume?, thanks I have included it now

I agree that the output format is confusing. We have reworked the whole metrics part in a "secret development version" of machamp (it now supports more detailed reporting of metrics), I will resolve this file format after merging that one into the main branch.

**Comment by robvanderg** [comment]
Solved in 0.4.2
